src/common/translations/localize.ts

```diff
@@ -42,7 +43,7 @@ export type FlattenObjectKeys<
 
 export type LocalizeFunc<Keys extends string = LocalizeKeys> = (
   key: Keys,
-  ...args: any[]
+  values?: Record<string, string | number | HTMLTemplateResult>
```

Comment:
Why are we allowing numbers and HTML templates when the output is still a string?  Stringification should be controlled where `localize` is called so the output is controlled per instance, not inside the method.

Reply:
I don't really get what you mean..?
FYI:
- it makes more sense to be able to pass numbers than to have to pass `String(theNumber)`
- it doesn't actually return a string all the time, if a template is passed it returns ["part a", html\`part b\`]

Reply:
> it makes more sense to be able to pass numbers than to have to pass `String(theNumber)`

I disagree.  There have been several display issues in the past stemming from number formatting (e.g. precision, alignment, internationalization, etc.).  A big benefit to better typing here is that it forces us to evaluate the possibilities for the number to get the formatting right the first time.  It won't always be just `toString()`.

> it doesn't actually return a string all the time, if a template is passed it returns ["part a", html`part b`]

😕 The return type is `string`.

Reply:
> There have been several display issues in the past stemming from number formatting (e.g. precision, alignment, internationalization, etc.)

But you can handle number formatting in the localized message. Plus if I didn't allow numbers, I would have to update even more code.

> 😕 The return type is string.

Yup -- there's no easy way to change that (at least I think - I'm going to test something in the TS playground real quick)

Reply:
Okay, so as for the return type being string, you could [make it be more finely typed](https://www.typescriptlang.org/play?#code/MYGwhgzhAEASAqBZAMvApgWwA7gC5ugG8BfaAejOhzGDQAsB7EAEzQCcAoDgSwDt82AMxoFkDYGBDcAXgUIdo0ABQBrNAE8AXNAi42fAOYAaaAyy5uDXhAD82gEppgDNswA8u-b2M69hgHwAlNqehgrKalq+Xj5mFlYQDk4u7qHeJmkG0AA+cEiomNT4QdpKmTl5KOjYeGiBANoAuhzEHM7WuNAg4pIyBAC8yqoaJnGW1tCDJIGT-kThbGi4AK5svNAARNS0jCzsGwDcLTOQ0GISUrJHbQmdAG4AjJNdPZdoSht03BsmhBu4X28G20GzAG2IMwo0TC7V00DuACZnt0Ln0Pl8fkR-oCDMDoLw0AB3SoFGpgfBKQIQ8iUMp+bwVBBVQq1BqNIA). Only problem is that sometimes we pass "localize("thing", {var: html`some html`})" to functions that only accept strings (eg the dialog shower). If we made that return the right type, we would probably just be reconverting it to `string` later so we can pass it to those functions.

Reply:
Hmm... can you save me some time and point me to those occurrences as well?  I'll take a look tomorrow morning.

Passing tag templates here in general is very bothersome as it frequently causes a11y issues that are a pain to solve, so if we can root some out here that would be great.  We unfortunately still have a lot of spots in the code where we hand a template to a string type.

Reply:
This should be all the cases:
```
     1  hassio/src/hassio-my-redirect.ts:79
     1  hassio/src/system/hassio-supervisor-info.ts:325
     3  src/common/integrations/protocolIntegrationPicked.ts:51
     1  src/components/ha-code-editor.ts:263
     1  src/components/ha-icon-picker.ts:32
     1  src/panels/config/blueprint/ha-blueprint-overview.ts:376
     1  src/panels/config/dashboard/ha-config-dashboard.ts:105
     1  src/panels/config/integrations/integration-panels/zwave_js/dialog-zwave_js-update-firmware-node.ts:332
```

Reply:
Sorry, finally got a chance too look at those, but I think that list is not correct?  Several are referencing files without any calls to `localize`.

Reply:
This is just the output from tsc, so it must be other TS errors.
(Also keep in mind that localize's return value changing in one place may affect another)